python/ray/util/worker_group.py

```diff
@@ -0,0 +1,308 @@
+        """Executes the input function and returns the output.
+
+        Args:
+            func (Callable): The function to execute.
+            args, kwargs: The arguments to pass into func.
+        """
+        return func(*args, **kwargs)
+
+
+@dataclass
+class WorkerMetadata:
+    """Metadata for each worker/actor.
+
+    This information is expected to stay the same throughout the lifetime of
+    actor.
+
+    Args:
+        node_id (str): ID of the node this worker is on.
+        node_ip (str): IP address of the node this worker is on.
+        hostname (str): Hostname that this worker is on.
+        gpu_ids (List[int]): List of CUDA IDs available to this worker.
+    """
+    node_id: str
+    node_ip: str
+    hostname: str
+    gpu_ids: Optional[List[int]]
+
+
+@dataclass
+class Worker:
+    """Class representing a Worker."""
+    actor: ActorHandle
+    metadata: WorkerMetadata
+
+
+def create_executable_class(executable_cls: Optional[Type] = None) -> Type:
+    """Create the executable class to use as the Ray actors."""
+    if not executable_cls:
+        return BaseWorkerMixin
+    elif issubclass(executable_cls, BaseWorkerMixin):
+        return executable_cls
+    else:
+
+        class _WrappedExecutable(executable_cls, BaseWorkerMixin):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+
+        return _WrappedExecutable
+
+
+def construct_metadata() -> WorkerMetadata:
+    """Creates metadata for this worker.
+
+    This function is expected to be run on the actor.
+    """
+    node_id = ray.get_runtime_context().node_id.hex()
+    node_ip = ray.util.get_node_ip_address()
+    hostname = socket.gethostname()
+    gpu_ids = ray.get_gpu_ids()
+
+    return WorkerMetadata(
+        node_id=node_id, node_ip=node_ip, hostname=hostname, gpu_ids=gpu_ids)
+
+
+class WorkerGroup:
+    """Group of Ray Actors that can execute arbitrary functions.
+
+    ``WorkerGroup`` launches Ray actors according to the given
+    specification. It can then execute arbitrary Python functions in each of
+    these workers.
+
+    If not enough resources are available to launch the actors, the Ray
+    cluster will automatically scale up if autoscaling is enabled.
+
+    Args:
+        num_workers (int): The number of workers (Ray actors) to launch.
+            Defaults to 1.
+        num_cpus_per_worker (float): The number of CPUs to reserve for each
+            worker. Fractional values are allowed. Defaults to 1.
+        num_gpus_per_worker (float): The number of GPUs to reserve for each
+            worker. Fractional values are allowed. Defaults to 0.
+        additional_resources_per_worker (Optional[Dict[str, float]]):
+            Dictionary specifying the extra resources that will be
+            requested for each worker in addition to ``num_cpus_per_worker``
+            and ``num_gpus_per_worker``.
+        actor_cls (Optional[Type]): If specified use this class as the
```

Comment:
For the general case, make `actor_cls` required and not Optional. Then on SGD-side implement a `SGDWorkerGroup` on top of this.